zio-flow/shared/src/main/scala/zio/flow/internal/DurableIndexedStore.scala

```diff
@@ -0,0 +1,120 @@
+  } yield position).mapError(s => new IOException(s))
+
+  def getNamespaces(): IO[IOException, Map[Chunk[Byte], ColumnFamilyHandle]] =
+    for {
+      list <- colHandles.get
+      map  <- ZIO.succeed(list.map(cHandle => Chunk.fromArray(cHandle.getName) -> cHandle).toMap)
+    } yield map
+
+  def getColFamilyHandle(topic: String): UIO[ColumnFamilyHandle] =
+    for {
+      namespaces <- getNamespaces().orDie
+      cf         <- ZIO.succeed(namespaces.get(ProtobufCodec.encode(Schema[String])(topic)))
+      handle <- cf match {
+                  case Some(h) => ZIO.succeed(h)
+                  case None    => addTopic(topic).orDie
+                }
+    } yield handle
+
+  override def put(topic: String, value: Chunk[Byte]): IO[IOException, Long] = for {
+    colFam <- getColFamilyHandle(topic)
+    _ <- transactionDB.atomically {
+           Transaction.getForUpdate(
+             colFam,
+             ProtobufCodec.encode(Schema[String])("POSITION").toArray,
+             exclusive = true
+           ) >>= { posBytes =>
+             Transaction
+               .put(
+                 colFam,
+                 ProtobufCodec.encode(Schema[String])("POSITION").toArray,
+                 incPosition(posBytes)
+               ) >>= { _ =>
+               Transaction
+                 .put(
+                   colFam,
+                   incPosition(posBytes),
+                   value.toArray
+                 )
+                 .refineToOrDie[IOException]
+             }
+           }
+         }.refineToOrDie[IOException]
+    newPos <- position(topic)
+  } yield newPos
+
+  private def incPosition(posBytes: Option[Array[Byte]]): Array[Byte] =
+    ProtobufCodec
+      .encode(Schema[Long])(
+        ProtobufCodec.decode(Schema[Long])(Chunk.fromArray(posBytes.get)) match {
+          case Left(error) => throw new IOException(error)
+          case Right(p)    => p + 1
+        }
+      )
+      .toArray
+
+  override def scan(topic: String, from: Long, inclusiveTo: Long): ZStream[Any, IOException, Chunk[Byte]] =
+    ZStream.fromEffect(getColFamilyHandle(topic)).flatMap { cf =>
+      for {
+        k <- ZStream.fromIterable(from to inclusiveTo)
+        value <-
+          ZStream
+            .fromEffect(transactionDB.get(cf, ProtobufCodec.encode(Schema[Long])(k).toArray).refineToOrDie[IOException])
+      } yield value.map(Chunk.fromArray).getOrElse(Chunk.empty)
+    }
+}
+
+object DurableIndexedStore {
+
+  def live(topic: String): ZLayer[Has[TransactionDB], Throwable, Has[DurableIndexedStore]] = (for {
+    colHandles    <- Ref.make(List.empty[ColumnFamilyHandle])
+    transactionDB <- ZIO.service[service.TransactionDB]
+    di            <- ZIO.succeed(DurableIndexedStore(transactionDB, colHandles))
+    _             <- di.addTopic(topic)
+  } yield di).toLayer
+
+  def live: ZLayer[Has[TransactionDB], Throwable, Has[DurableIndexedStore]] = (for {
```

Comment:
You could extract the constructor effect and reuse in both `live`s to avoid redundancy